core/src/main/java/brooklyn/management/entitlement/PerUserEntitlementManager.java

```diff
@@ -0,0 +1,97 @@
+import brooklyn.entity.basic.ConfigKeys;
+import brooklyn.util.collections.MutableMap;
+import brooklyn.util.text.Strings;
+
+import com.google.common.base.Preconditions;
+
+public class PerUserEntitlementManager implements EntitlementManager {
+
+    private static final Logger log = LoggerFactory.getLogger(PerUserEntitlementManager.class);
+    
+    public final static String PER_USER_ENTITLEMENTS_CONFIG_PREFIX = Entitlements.ENTITLEMENTS_CONFIG_PREFIX+".perUser";
+    
+    public final static ConfigKey<String> DEFAULT_MANAGER = ConfigKeys.newStringConfigKey(PER_USER_ENTITLEMENTS_CONFIG_PREFIX+
+        ".default", "Default entitlements manager for users without further specification", "minimal");
+    
+    protected final EntitlementManager defaultManager;
+    protected final Map<String,EntitlementManager> perUserManagers = MutableMap.of();
+
+    private final static ThreadLocal<Boolean> ACTIVE = new ThreadLocal<Boolean>();
+    
+    private static EntitlementManager load(BrooklynProperties properties, String type) {
+        if (Boolean.TRUE.equals(ACTIVE.get())) {
+            // prevent infinite loop
+            throw new IllegalStateException("Cannot set "+PerUserEntitlementManager.class.getName()+" within config for itself");
+        }
+        try {
+            ACTIVE.set(true);
+            return Entitlements.load(null, properties, type);
+        } finally {
+            ACTIVE.remove();
+        }
+    }
+    
+    public PerUserEntitlementManager(BrooklynProperties properties) {
+        this(load(properties, properties.getConfig(DEFAULT_MANAGER)));
+        
+        BrooklynProperties users = properties.submap(ConfigPredicates.startingWith(PER_USER_ENTITLEMENTS_CONFIG_PREFIX+"."));
+        for (Map.Entry<ConfigKey<?>,?> key: users.getAllConfig().entrySet()) {
+            if (key.getKey().getName().equals(DEFAULT_MANAGER.getName())) continue;
+            String user = Strings.removeFromStart(key.getKey().getName(), PER_USER_ENTITLEMENTS_CONFIG_PREFIX+".");
+            addUser(user, load(properties, Strings.toString(key.getValue())));
+        }
+        
+        log.info(getClass().getSimpleName()+" created with "+perUserManagers.size()+" user"+Strings.s(perUserManagers)+" and "
+            + "default "+defaultManager+" (users: "+perUserManagers+")");
+    }
+    
+    public PerUserEntitlementManager(EntitlementManager defaultManager) {
+        this.defaultManager = Preconditions.checkNotNull(defaultManager);
+    }
+
+    public void addUser(String user, EntitlementManager managerForThisUser) {
+        perUserManagers.put(user, managerForThisUser);
+    }
+
+    @Override
+    public <T> boolean isEntitled(EntitlementContext context, EntitlementClass<T> entitlementClass, T entitlementClassArgument) {
+        EntitlementManager entitlementInEffect = null;
+        if (context==null || context.user()==null) {
+            // no user means it is running as an internal process, always has root
```

Comment:
I'd prefer if we could use some special context to indicate this, rather than null. Can we intercept earlier, e.g. in `Entitlements.getEntitlementContext()` then don't return null.

The reason is that treating null as super-user feels risky as the behaviour if there is a programming error which would otherwise cause a NPE!